source/ceylon/toml/formatToml.ceylon

```diff
@@ -0,0 +1,269 @@
+ that the argument be a [[TomlTable]], although any `Map` with valid content can be used.
+
+ Note that all keys must be [[String]]s, and all values must be [[TomlValue]]s, [[Map]]s,
+ or [[List]]s. No type coercion is performed by this function."
+throws(`class AssertionError`, "if the argument represents an invalid TOML document")
+shared String formatToml(Map<String, Object> tomlTable)
+        => object satisfies Producer<String> {
+
+    value tablePath = ArrayList<String>();
+
+    value sb = StringBuilder();
+
+    AssertionError error(String message) {
+        value location
+            =   if (tablePath.empty)
+                then ""
+                else " at key path '``formatKeyPath()``'";
+        return AssertionError(message + location);
+    }
+
+    Boolean isArrayOfTables(Anything item)
+        =>  if (!is List<Anything> item)
+            then false
+            else item.first is Map<Anything, Anything>;
+
+    Boolean isSimpleValue(Anything item)
+        =>     !item is Map<Anything, Anything>
+            && !isArrayOfTables(item);
+
+    void emitNewline()
+        =>  sb.append(operatingSystem.newline);
+
+    String escapeKey(String key) {
+        if (key.every(isValidKeyCharacter)) {
+            return key;
+        }
+        value sb = StringBuilder();
+        emitBasicString(key, sb);
+        return sb.string;
+    }
+
+    void emitKeyPath(StringBuilder sb = this.sb) {
+        variable value first = true;
+        for (path in tablePath) {
+            if (!first) {
+                sb.appendCharacter('.');
+            }
+            else {
+                first = false;
+            }
+            sb.append(path);
+        }
+    }
+
+    String formatKeyPath() {
+        value sb = StringBuilder();
+        emitKeyPath(sb);
+        return sb.string;
+    }
+
+    void emitTableHeading(Boolean isArray = false) {
+        if (!tablePath.empty) {
+            if (!sb.empty) {
+                emitNewline();
+            }
+            sb.append(isArray then "[[" else "[");
+            emitKeyPath();
+            sb.append(isArray then "]]" else "]");
+            emitNewline();
+        }
+    }
+
+    void emitStringValue(String item) {
+        // TODO analyze item and choose best encoding type?
+        //      (multiline & literal combos)
+        emitBasicString(item);
+    }
+
+    void emitBasicString(String item, StringBuilder sb = this.sb) {
+        sb.appendCharacter('"');
+        for (char in item) {
+            switch (char)
+            case ('\{#08}') { sb.append("\\b"); }
+            case ('\{#09}') { sb.append("\\t"); }
+            case ('\{#0a}') { sb.append("\\n"); }
+            case ('\{#0c}') { sb.append("\\f"); }
+            case ('\{#0d}') { sb.append("\\r"); }
+            case ('\{#22}') { sb.append("\\\""); }
+            case ('\{#5c}') { sb.append("\\\\"); }
+            else if (char < '\{#20}') {
```

Comment:
What about U+007F DEL? And I think there might be some more undesirable Unicode characters.